lib/ringsc.gi

```diff
@@ -403,14 +402,58 @@ InstallGlobalFunction( RingByStructureConstants, function( arg )
         NewType( Fam, IsSCRingObj and IsDenseCoeffVectorRep );
 
     SetCoefficientsFamily( Fam, ElementsFamily( FamilyObj( Integers ) ) );
-    # temporary
-    SetIsUFDFamily(Fam,false);
 
     # Make the generators and the ring.
     SetZero( Fam, ObjByExtRep( Fam, List( [ 1 .. n ], x -> 0 ) ) );
     gens:= Immutable( List( IdentityMat( n, Integers ),
                             x -> ObjByExtRep( Fam, x ) ) );
-    A:= RingByGenerators( gens );
+
+    # Hope to find a unit (and thus an identity)
+    one:=fail;
+    if oneinfo=true then
+      # find one, if possible
+      # Try generators first
+      i:=1;
+      while one=fail and i<=Length(gens) do
+        A:=[];
+        new:=gens[i];
+        repeat
+          Add(A,new);
+          new:=new*gens[i];
+        until new in A;
```

Comment:
We need the whole list (to find a loop. Of course one could use the idea of Pollard rho to avoid the list.
It is a very basic implementation that can handle the cases of smallish rings (which is all I needed for the time being), and I hope will at some point be replaced with a better method (solving equations).